File: androld/app/src/main/java/z3d/res/RoleRes.java
package z3d.res;

import android.util.Log;

import java.util.ArrayList;
import java.util.List;

import z3d.base.ByteArray;

import z3d.base.CallBackFun;
import z3d.base.RoleBackFun;
import z3d.filemodel.MeshDataManager;
import z3d.units.AnimManager;
import z3d.vo.SkinMesh;
import z3d.vo.Vector3D;

public class RoleRes extends BaseRes {
    public String roleUrl;
    public List<String> actionAry ;
    private RoleBackFun bfun;
    public int meshBatchNum;
    public Vector3D ambientLightColor;
    public float ambientLightIntensity;
    public Vector3D sunLigthColor;
    public float sunLigthIntensity;
    public Vector3D nrmDircet;


    public void  loadComplete(byte[] buff, RoleBackFun bfun)
    {
        this._byte =new ByteArray(buff);
        this.version = this._byte.readInt();

        this.readMesh();
        this.readAction();
        this.read(new CallBackFun() {
            @Override
            public void StateChange(boolean State) {
                readNext();
            }
        });;//readimg

    }
    private void   readNext() {


    }
    public void readMesh() {
        this.roleUrl = this._byte.readUTF();
        if (this.version >= 16) { //环境参数
            this.ambientLightColor = new Vector3D();
            this.sunLigthColor = new Vector3D();
            this.nrmDircet = new Vector3D();

            this.ambientLightColor.x = this._byte.readFloat();
            this.ambientLightColor.y = this._byte.readFloat();
            this.ambientLightColor.z = this._byte.readFloat();
            this.ambientLightIntensity = this._byte.readFloat();
            this.ambientLightColor.scaleBy(this.ambientLightIntensity);

            this.sunLigthColor.x = this._byte.readFloat();
            this.sunLigthColor.y = this._byte.readFloat();
            this.sunLigthColor.z = this._byte.readFloat();
            this.sunLigthIntensity = this._byte.readFloat();
            this.sunLigthColor.scaleBy(this.sunLigthIntensity);

            this.nrmDircet.x = this._byte.readFloat();
            this.nrmDircet.y = this._byte.readFloat();
            this.nrmDircet.z = this._byte.readFloat();
        }

          MeshDataManager.getInstance().readData(this._byte, this.meshBatchNum, this.roleUrl, this.version);


    }
    private void readAction() {
        ByteArray $actionByte;
        if (this.version >= 30) {
            $actionByte = this.getZipByte(this._byte);
        } else {
            $actionByte = this._byte;
        }
        this.actionAry = new ArrayList<>();
        int actionNum = $actionByte.readInt();
        for (int i = 0; i < actionNum; i++) {
            String actionName = $actionByte.readUTF();
            Log.d(actionName, "actionName: ");
            AnimManager.getInstance().readData($actionByte, this.roleUrl + actionName);
            this.actionAry.add(actionName);
        }


    }

}

File: metalios/iosgl/iosgl/bt/res/RoleRes.m
//
//  RoleRes.m
//  iosgl
//
//  Created by zhao on 2/4/2020.
//  Copyright © 2020 zhao. All rights reserved.
//

#import "RoleRes.h"
#import "Vector3D.h"
#import "GL_Header.h"
#import "ByteArray.h"
#import "AnimManager.h"
#import "LoadManager.h"
#import "MeshDataManager.h"

 
@interface RoleRes()


@property(nonatomic,strong)SuccessBlock fun;
@property(nonatomic,strong)Vector3D* ambientLightColor;
@property(nonatomic,strong)Vector3D* sunLigthColor;
@property(nonatomic,strong)Vector3D* nrmDircet;

@property(nonatomic,assign)float ambientLightIntensity;
@property(nonatomic,assign)float sunLigthIntensity;


@end

@implementation RoleRes


-(void)load:(NSString*)url  fun:(SuccessBlock)fun;
{
    self.fun=fun;
    [[LoadManager default] loadUrl: url type:LoadManager.BYTE_TYPE fun:^(NSString* value) {
        NSDictionary* dic=(NSDictionary*)value;
        NSData* netNsData = [[NSData alloc] initWithContentsOfFile:dic[@"data"]];
       
        [self loadComplete:[[ByteArray alloc]init:netNsData]];
    }];
 
}
-(void)loadComplete:(ByteArray*)byte;
{
    self.byte=byte;
    self.version= [self.byte readInt];
    [self readMesh];
    [self readAction];
 
}
-(void)readAction;
{
    ByteArray *actionByte;
    if(self.version>=30){
        int zipLen = [self.byte readInt];
        NSData *zipNsData=  [self.byte getNsDataByLen:zipLen];
        NSData *outputData =[self gzipInflate:zipNsData] ;
        NSLog(@"len-%d-解压后长度>%ld",zipLen,outputData.length);
        actionByte=  [[ByteArray alloc]init:outputData];
    }else{
        actionByte=self.byte;
    }
    
    self.actionAry=[[NSMutableArray alloc]init];
    int actionNum = [actionByte readInt];
    for (int i = 0; i < actionNum; i++) {
        NSString* actionName   = [actionByte readUTF];
        [self.scene3D.animManager readData:actionByte url:[self.roleUrl stringByAppendingString:actionName]];
        [self.actionAry addObject:actionName];
    }
    [self read:^(NSString* code) {
        [self readNext];
    }];
  
}
-(void) readNext;
{
    [self read];
    [self read];
    self.fun(@"1");
}
-(void)readMesh;
{
    RoleRes* this=self;
    this.roleUrl = [this.byte readUTF];
    this.ambientLightColor = [Vector3D alloc];
    this.sunLigthColor = [Vector3D alloc];
    this.nrmDircet = [Vector3D alloc];
    
    
    this.ambientLightColor.x = [this.byte readFloat];
    this.ambientLightColor.y = [this.byte readFloat];
    this.ambientLightColor.z = [this.byte readFloat];
    this.ambientLightIntensity = [this.byte readFloat];
    [this.ambientLightColor scaleBy:this.ambientLightIntensity];

    this.sunLigthColor.x = [this.byte readFloat];
    this.sunLigthColor.y = [this.byte readFloat];
    this.sunLigthColor.z = [this.byte readFloat];
    this.sunLigthIntensity = [this.byte readFloat];
    [this.sunLigthColor scaleBy:this.sunLigthIntensity];

    this.nrmDircet.x = [this.byte readFloat];
    this.nrmDircet.y = [this.byte readFloat];
    this.nrmDircet.z = [this.byte readFloat];
    

    
    [    this.scene3D.meshDataManager readData:self.byte batchNum:this.meshBatchNum url:this.roleUrl version:this.version];
}
 
@end
